### PR TITLE
Add location to type matching errors

### LINKED ISSUE #268: Report error in def body only
Currently typing errors are shown as a problem of the expression - which usually means the whole file.

Good error:

Bad error:

In principle we should be able to report the error in the body of definition where the error occurred. The previous `def`s, `let`s and commands must have been fine as we checked them. They could of course lead to the failure in some later part, but only indirectly.

At minimum we should be able to split at `def`s as those must be at global scope and each individually typecheck.

### PR DESCRIPTION
**Pull request by xsebek**
- closes #268 
- computes the location for `SBind` and operators

### CODE REVIEW

**Comment by xsebek** [comment]
There might be other escape routes or a better way to annotate the error, but his seems to help. :man_shrugging:

**Comment by byorgey** [comment]
This looks OK, but is there any reason not to add a single call to `addLocToTypeError` at the very top of `inferModule`?  As it is, it will only catch any errors that happen specifically on the two lines where you added it.

**Comment by xsebek** [comment]
@byorgey nope, I was just afraid you would beat me to it, so I pushed the first thing that worked :smile:

**Comment by byorgey** [comment]
> nope, I was just afraid you would beat me to it

lol, I was actually working on it when you pushed this, so well done :laughing:

**Review by byorgey — APPROVED** [review]

File: src/Swarm/Language/Parse.hs [hunk 1]
```diff
@@ -329,13 +329,13 @@ parseExpr = fixDefMissingSemis <$> makeExprParser parseTermAtom table
 exprLoc2 :: Parser (Syntax -> Syntax -> Term) -> Parser (Syntax -> Syntax -> Syntax)
```
Comment (byorgey) [inline src/Swarm/Language/Parse.hs:329]: We should update the comment too.

File: src/Swarm/Language/Typecheck.hs [hunk 2]
```diff
@@ -280,8 +280,10 @@ inferModule s@(Syntax _ t) = case t of
   -- correct context when checking the right-hand side in particular.
   SBind mx c1 c2 -> do
     -- First, infer the left side.
-    Module cmda ctx1 <- inferModule c1
-    a <- decomposeCmdTy cmda
+    (a, ctx1) <- (`catchError` addLocToTypeErr c1) $ do
```
Comment (byorgey) [inline src/Swarm/Language/Typecheck.hs:283]: Why do we need this here?  Just curious.

Comment (xsebek) [reply src/Swarm/Language/Typecheck.hs:283]: Oh, I was just going crazy from bind chains - turns out they were missing location themselves and could not help to limit the error scope. I will make this simple again :+1:

**Comment by xsebek** [comment]
The error scope for type mismatch seems to be reasonable now. :partying_face:
But I better add some unit tests now that it is no longer a two line change :sweat_smile:

File: src/Swarm/Language/Parse.hs (hunk 1 above)
Comment (xsebek) [reply src/Swarm/Language/Parse.hs]: Done and I made it a local definition too :+1:

**Comment by byorgey** [comment]
LGTM!

**Pull request merged**